classes/local/store/object_file_system.php

```diff
@@ -736,8 +736,39 @@ protected function copy_from_external_to_local($contenthash) {
      * @throws \dml_exception
      */
     public function redirect_to_presigned_url($contenthash, $headers = array()) {
+        global $FULLME;
         try {
-            redirect($this->externalclient->generate_presigned_url($contenthash, $headers));
+            $signedurl = $this->externalclient->generate_presigned_url($contenthash, $headers);
+            if (headers_sent()) {
+                debugging('objectfs redirect for ' . $contenthash . ' from ' . $FULLME .
+                        ': headers already sent; redirect may be incorrectly cached in browser');
+            } else {
+                // Remove all previously-set headers, and look for cache-control setting.
+                $cachecontrol = '';
+                foreach (headers_list() as $header) {
+                    // Get header name (text before the colon).
+                    if (preg_match('~^([^:]+):(.*)$~', $header, $matches)) {
+                        if (strtolower($matches[1]) === 'cache-control') {
+                            $cachecontrol = $matches[2];
+                        }
+                        header_remove($matches[1]);
+                    }
+                }
+                // Set expires and cache-control values to match the presigned URL expiry, which may be
+                // different to values previously set.
+                header('Expires: '. gmdate('D, d M Y H:i:s', $signedurl->expiresat) .' GMT');
+                // Unless cache-control was previously set to 'public' by Moodle for the actual file send,
+                // use 'private' to allow browser caching only; otherwise via a shared cache users might
+                // be able to redirect to content that was only supposed to be displayed to a different
+                // user.
+                if (strpos($cachecontrol, 'public') !== false) {
+                    $publicprivate = 'public';
+                } else {
+                    $publicprivate = 'private';
+                }
```

Comment:
```suggestion
                $publicprivate = 'private';
                if (strpos($cachecontrol, 'public') !== false) {
                    $publicprivate = 'public';
                }
```
I prefer less `else` blocks (reduces mental overhead). This would achieve more or less the same output and reads better personally.

Reply:
Also could you change the variable name to `$cachevisibility` instead?